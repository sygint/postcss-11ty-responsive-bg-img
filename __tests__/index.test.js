const { rimrafSync } = require("rimraf");
const postcss = require("postcss");

const plugin = require("../");

const imgUrl = "./__tests__/images/happy-cat.png";
const singleQuoteImgUrl = './__tests__/images/happy-ca\'t.png';
const doubleQuoteImgUrl = './__tests__/images/happy-ca\"t.png';
const singleWidth = "500";
const multipleWidths = "400, 640, 768";

const errorCases = {
  "no params": [
    ".no-params-error { background: responsive-bg-img() }",
    "Missed required url and width parameters",
  ],
  "missing url parameter": [
    `.missing-url-error { background: responsive-bg-img(${singleWidth}) }`,
    "Missed required url parameter",
  ],
  "missing width parameter": [
    `.missing-width-error { background: responsive-bg-img(${imgUrl}) }`,
    "Missed required width parameter",
  ],
  "incorrect url parameter": [
    `.incorrect-url-error { background: responsive-bg-img(${imgUrl}.invalid, ${singleWidth}) }`,
    "Invalid filename: ./__tests__/images/happy-cat.png.invalid",
  ],
};

const cases = {
  "background-image rule": `.background-image-test { background-image: responsive-bg-img(${imgUrl}, ${singleWidth}) }`,
  "single quotes for url": `.single-quote-test { background: responsive-bg-img('${imgUrl}', ${singleWidth}) }`,
  "double quotes for url": `.double-quote-test { background: responsive-bg-img("${imgUrl}", ${singleWidth}) }`,
  "single quotes for double quote url": `.single-quote-double-quote-test { background: responsive-bg-img('${doubleQuoteImgUrl}', ${singleWidth}) }`,
  "double quotes for single quote url": `.double-quote-single-quote-test { background: responsive-bg-img("${singleQuoteImgUrl}", ${singleWidth}) }`,
  "multiple widths": `.multiple-widths-test { background: responsive-bg-img(${imgUrl}, ${multipleWidths}) }`,
  "url function": `.url-function-test { background: responsive-bg-img(url(${imgUrl}), ${multipleWidths}) }`,
  "url function with single quotes": `.url-function-single-quotes-test { background: responsive-bg-img(url('${imgUrl}'), ${singleWidth}) }`,
  "url function with double quotes": `.url-function-double-quotes-test { background: responsive-bg-img(url("${imgUrl}"), ${singleWidth}) }`,
};

async function test(input, opts = {}) {
  const result = await postcss([plugin(opts)]).process(input, {
    from: undefined,
  });

  expect(result.css).toMatchSnapshot();
  expect(result.warnings()).toHaveLength(0);
}

async function testError(input, error) {
  const result = async () =>
    await postcss([plugin]).process(input, { from: undefined });
  await expect(result).rejects.toThrow(error);
}

describe("postcss-11ty-responsive-bg-img", () => {
  afterAll(() => {
    rimrafSync("./img/");
  });

  for (let errorCase in errorCases) {
    const [input, errorText] = errorCases[errorCase];
    const description = `throws "${errorText}" error when called with ${errorCase}`;
    it(description, async () => await testError(input, errorText));
  }

  for (let caseName in cases) {
    const input = cases[caseName];
    const description = `renders correctly when called with ${caseName}`;
    it(description, async () => await test(input));
  }
});
