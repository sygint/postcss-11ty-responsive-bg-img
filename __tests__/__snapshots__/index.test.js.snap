// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcss-11ty-responsive-bg-img renders correctly when called with background-image rule 1`] = `".background-image-test { background-image: url(\\"/img/HOdUYjZ3EM-500.webp\\") }"`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with double quotes for url 1`] = `".double-quote-test { background: url(\\"/img/HOdUYjZ3EM-500.webp\\") }"`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with multiple widths 1`] = `
".multiple-widths-test { background: url(\\"/img/HOdUYjZ3EM-400.webp\\") }
@media (min-width: 640px) { .multiple-widths-test { background-image: url(\\"/img/HOdUYjZ3EM-640.webp\\"); } }
@media (min-width: 768px) { .multiple-widths-test { background-image: url(\\"/img/HOdUYjZ3EM-768.webp\\"); } }"
`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with single quotes for url 1`] = `".single-quote-test { background: url(\\"/img/HOdUYjZ3EM-500.webp\\") }"`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with url function 1`] = `
".url-function-test { background: url(\\"/img/HOdUYjZ3EM-400.webp\\") }
@media (min-width: 640px) { .url-function-test { background-image: url(\\"/img/HOdUYjZ3EM-640.webp\\"); } }
@media (min-width: 768px) { .url-function-test { background-image: url(\\"/img/HOdUYjZ3EM-768.webp\\"); } }"
`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with url function with double quotes 1`] = `".url-function-double-quotes-test { background: url(\\"/img/HOdUYjZ3EM-500.webp\\") }"`;

exports[`postcss-11ty-responsive-bg-img renders correctly when called with url function with single quotes 1`] = `".url-function-single-quotes-test { background: url(\\"/img/HOdUYjZ3EM-500.webp\\") }"`;
